src/main/java/org/testng/internal/TestNgMethodUtils.java

```diff
@@ -47,7 +50,87 @@ private static boolean isConfigurationMethod(ITestNGMethod method, boolean inclu
      * @param methods - A List of {@link ITestNGMethod} in which the check needs to be done.
      * @return - <code>true</code> if the method is a configuration method and exists in the list of methods passed.
      */
-    static boolean containsConfigurationMethod(ITestNGMethod method, List<ITestNGMethod> methods) {
+    private static boolean containsConfigurationMethod(ITestNGMethod method, List<ITestNGMethod> methods) {
         return isConfigurationMethod(method, true) && methods.contains(method);
     }
+
+    static ITestNGMethod[] filterBeforeTestMethods(ITestClass testClass, Invoker.Predicate<ITestNGMethod, IClass> predicate) {
+        return filterMethods(testClass, testClass.getBeforeTestMethods(), predicate);
+    }
+
+    static ITestNGMethod[] filterAfterTestMethods(ITestClass testClass, Invoker.Predicate<ITestNGMethod, IClass> predicate) {
+        return filterMethods(testClass, testClass.getAfterTestMethods(), predicate);
+    }
+
+    /**
+     * @return Only the ITestNGMethods applicable for this testClass
+     */
+    static ITestNGMethod[] filterMethods(IClass testClass,
+                                         ITestNGMethod[] methods,
+                                         Invoker.Predicate<ITestNGMethod, IClass> predicate) {
+        List<ITestNGMethod> vResult= Lists.newArrayList();
+
+        for(ITestNGMethod tm : methods) {
+            String msg;
+            if (predicate.isTrue(tm, testClass) &&  (!TestNgMethodUtils.containsConfigurationMethod(tm, vResult)) ) {
+                msg =  "Keeping method " + tm + " for class " + testClass;
+                vResult.add(tm);
+            } else {
+                msg = "Filtering out method " + tm + " for class " + testClass;
+            }
+            Utils.log("Invoker " + Thread.currentThread().hashCode(), 10, msg);
+        }
+        return vResult.toArray(new ITestNGMethod[vResult.size()]);
+    }
+
+    /**
+     * The array of methods contains @BeforeMethods if isBefore if true, @AfterMethods
+     * otherwise.  This function removes all the methods that should not be run at this
+     * point because they are either firstTimeOnly or lastTimeOnly and we haven't reached
+     * the current invocationCount yet
+     */
+    static ITestNGMethod[] filterFirstTimeRunnableSetupConfigurationMethods(ITestNGMethod tm, ITestNGMethod[] methods) {
```

Comment:
Maybe we can use collections instead of arrays

Reply:
That would cause a lot of other changes to be triggered.

Reply:
ok